src/ApolloHooksMutation.re

```diff
@@ -0,0 +1,210 @@
+type graphqlErrors;
+
+type error = {
+  .
+  "message": string,
+  "graphqlErrors": graphqlErrors,
+};
+
+type refetchQueries =
+  ReasonApolloTypes.executionResult => array(ApolloClient.queryObj);
+
+/* Result that is return by the hook */
+type result('a) =
+  | Data('a)
+  | Error(error)
+  | NoData;
+
+/* Result that is return by the hook */
+type controlledResult('a) = {
+  loading: bool,
+  called: bool,
+  data: option('a),
+  error: option(error),
+};
+
+type controlledVariantResult('a) =
+  | Loading
+  | Called
+  | Data('a)
+  | Error(error)
+  | NoData;
+
+[@bs.module "graphql-tag"] external gql: ReasonApolloTypes.gql = "default";
+
+type mutationResult('a) = {. "data": 'a};
+
+[@bs.deriving abstract]
+type options('a) = {
+  [@bs.optional]
+  variables: Js.Json.t,
+  [@bs.optional]
+  client: ApolloClient.generatedApolloClient,
+  [@bs.optional]
+  refetchQueries,
+  [@bs.optional]
+  awaitRefetchQueries: bool,
+  [@bs.optional]
+  update: (ApolloClient.generatedApolloClient, mutationResult('a)) => unit,
+};
+
+type jsResult = {
+  .
+  "data": Js.Nullable.t(Js.Json.t),
+  "loading": bool,
+  "called": bool,
+  "error": Js.Nullable.t(error),
+};
+
+type jsMutate('a) = (. options('a)) => Js.Promise.t(jsResult);
+type mutation('a, 'b) =
+  (
+    ~mutation: 'b,
+    ~client: ApolloClient.generatedApolloClient=?,
+    ~refetchQueries: refetchQueries=?,
+    ~awaitRefetchQueries: bool=?,
+    unit
+  ) =>
+  Js.Promise.t(controlledVariantResult('a));
+
+[@bs.module "@apollo/react-hooks"]
+external useMutationJs:
+  (. ReasonApolloTypes.queryString, options('a)) => (jsMutate('a), jsResult) =
+  "useMutation";
+
+exception Error(string);
+
+let useMutation =
+    (
+      ~incompleteMutation=?,
```

Comment:
I am not sure `incompleteMutation` is the right name here, it might get really verbose and tedious to write on many mutations, can we come up with something more succinct for this case? 😄

Reply:
Usually you would use the mutation argument, unless you need to set the arguments dynamically. It would be better to have the mutation as an optional argument, but that needs some changes on the javascript side (I think!). Anyway I agree it's not ideal (and hopefully temporarily). Let me know what better name you are thinking of!